### ISSUE TITLE
dyld: Library not loaded: libdb2.dylib when running GO ibm_DB sdk in MAC

### ISSUE THREAD

**Issue by sanjeevghimire** [ibmdb/go_ibm_db#24]
when running the code snippet I get this error:

I installed the driver as per the instruction in the readme. the lib folder has the file libdb2.dylib but it complains about it:

```dyld: Library not loaded: libdb2.dylib
  Referenced from: /var/folders/_h/4trwzcd908g703_vdw2q1yn00000gn/T/go-
build774492263/b001/exe/db2conn
  Reason: image not found
signal: abort trap```

the code I am using is:

```
package main

import (
    _ "github.com/ibmdb/go_ibm_db"
    "database/sql"
    "fmt"
)

func main(){
    con:="HOSTNAME=host;DATABASE=db;PORT=50000;UID=u;PWD=pwd"
 db, err:=sql.Open("go_ibm_db", con)
    if err != nil{
        
		fmt.Println(err)
	}
	db.Close()
}

```

**Comment by sanjeevghimire** [comment]
Figured out the problem. My path was pointing to wrong DYLD library.

**Comment by bombsimon** [comment]
@akhilravuri1 I still cannot get this to work. I installed the driver with `setup.go`. The solution in this issue and #50 and #47 and #29 all seem to be resolved with the correct environment. Any help would be appreciated!

This is my environemnt:

```sh
$ for env in $DB2HOME $CGO_CFLAGS $CGO_LDFLAGS $DYLD_LIBRARY_PATH; do echo $env; done
/Users/simon/tmp/db2test/clidriver
-I/Users/simon/tmp/db2test/clidriver/include
-L/Users/simon/tmp/db2test/clidriver/lib
/Users/simon/tmp/db2test/clidriver/lib
```

When trying to use the driver, I get the error form this issue:

```sh
$ go run main.go
dyld: Library not loaded: libdb2.dylib
  Referenced from: /var/folders/q3/bf76j42160j2nt4s2tz1_fd40000gn/T/go-build982956092/b001/exe/main
  Reason: image not found
signal: abort trap
```

However, the file mentioned seems to be present according to me:

```sh
$ stat -f "%N (%z)" $DYLD_LIBRARY_PATH/libdb2.dylib
/Users/simon/tmp/db2test/clidriver/lib/libdb2.dylib (32902420)
```

What am I missing here? Why does it seem like I'm missing the library when trying to use the driver?

**Comment by bombsimon** [comment]
Ok this was new to me. Don't know what this is or why but it seems like this only occurs with `go run` (and the executable is being ran from a temp folder). If I compile the program first it works as expected.

```sh
$ go run main.go; echo $?                                                                                                                                                                                                                                                                                                 
dyld: Library not loaded: libdb2.dylib
  Referenced from: /var/folders/q3/bf76j42160j2nt4s2tz1_fd40000gn/T/go-build556855419/b001/exe/main
  Reason: image not found
signal: abort trap
1
```

```sh
$ go build main.go && ./main; echo $?
0
```

Is this because the driver requires a relative path? It sounds odd because `libdb2.dylib` isn't relative to `main.go` but that's the only guess I have. 🤔

**Comment by kierquebs** [comment]
Try this one

instead of:
`go run main.go`

I used:
`go build -v`

then to your command line
`./project_name`

**Comment by akhilravuri1** [comment]
Hi @bombsimon 

please provide your mac version and go version.

Thanks,
Akhil

**Comment by bombsimon** [comment]
@kierquebs Thanks! Like I wrote in my last post in March it's working if I first build the binary and then run it so the question is why that is.

@akhilravuri1 I'm running macOS Catalina 10.15.4 with go version go1.14 darwin/amd64.

This is no major issue but I was just curious why it's possible to build and execute but not perform `go run`.

**Comment by robherley** [comment]
Hey I know this issue is kinda old but @bombsimon to answer your question or anyone else who finds this, the reason is here: https://github.com/golang/go/issues/36572

As for the workaround, you can just use `go run -exec` like so:

```
go run -exec "env DYLD_LIBRARY_PATH=path/to/your/lib" main.go
```

**Comment by htdinh13** [comment]
Beside the workaround with `go run -exec`, do we have permanent fix for that issue?

**Comment by Anjali-Saha** [comment]
I would also like to know if there is a permanent fix to this issue?  I get this error with 1.15 & 1.16 on mac 11.4 the only version that does not return this is go 1.13

```
dyld: Library not loaded: iclis09b.dylib
  Referenced from: /var/folders/w6/kzlc3_4n7f19wyl7r11mwk1r0000gq/T/go-build346883109/b001/exe/main
  Reason: image not found
signal: abort trap
```

```
twmbp4026:tw-bs-cit asaha$ go env
GO111MODULE="off"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/asaha/Library/Caches/go-build"
GOENV="/Users/asaha/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/asaha/Development/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/asaha/Development/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/w6/kzlc3_4n7f19wyl7r11mwk1r0000gq/T/go-build370582690=/tmp/go-build -gno-record-gcc-switches -fno-common"
```
